openforcefield/typing/engines/smirnoff/parameters.py

```diff
@@ -555,15 +646,112 @@ def __init__(self, allow_cosmetic_attributes=False, **kwargs):
+                raise SMIRNOFFSpecError(msg)
+
+    def _process_indexed_mapped_attributes(self, smirnoff_data):
+        # TODO: construct data structure for holding indexed_mapped attrs, which
+        # will get fed into setattr
+        indexed_mapped_attr_lengths = {}
+        reindex = set()
+        reverse = defaultdict(dict)
+
+        kwargs = list(smirnoff_data.keys())
+        for kwarg in kwargs:
+            attr_name, index, key = self._split_attribute_index_mapping(kwarg)
+            
+            # Check if this is an indexed_mapped attribute.
+            if ((key is not None) and
+                (index is not None) and
+                attr_name in self._get_indexed_mapped_parameter_attributes()):
+                
+                # we start with a dict because have no guarantee of order
+                # in which we will see each kwarg
+                # we'll switch this to a list later
+                if attr_name not in smirnoff_data:
+                    smirnoff_data[attr_name] = dict()
+                    reindex.add(attr_name)
+
+                if index not in smirnoff_data[attr_name]:
+                    smirnoff_data[attr_name][index] = dict()
+
+                smirnoff_data[attr_name][index][key] = smirnoff_data[kwarg]
+                del smirnoff_data[kwarg]
+
+                # build reverse mapping; needed for contiguity check below
+                if index not in reverse[attr_name]:
+                    reverse[attr_name][index] = dict()
+                reverse[attr_name][index][key] = kwarg
+
+        # turn all our top-level dicts into lists
+        # catch cases where we skip an index,
+        # e.g. k1_bondorder*, k3_bondorder* defined, but not k2_bondorder*
+        for attr_name in reindex:
+            indexed_mapping = []
+            j = 0
+            for i in sorted(smirnoff_data[attr_name].keys()):
+                if int(i) == j:
+                    indexed_mapping.append(smirnoff_data[attr_name][i])
+                    j += 1
+                else:
+                    # any key will do; we are sensitive only to top-level index
+                    key = sorted(reverse[attr_name][i].keys())[0]
+                    kwarg = reverse[attr_name][i][key]
+                    val = smirnoff_data[attr_name][i][key]
+
+                    msg = (f"Unexpected kwarg ({kwarg}: {val})  passed to {self.__class__} constructor. "
+                            "If this is a desired cosmetic attribute, consider setting "
+                            "'allow_cosmetic_attributes=True'")
+                    raise SMIRNOFFSpecError(msg)
+
+            smirnoff_data[attr_name] = indexed_mapping
+
+            # keep track of lengths; used downstream for checking against other
+            # indexed attributes
+            indexed_mapped_attr_lengths[attr_name] = len(smirnoff_data[attr_name])
+
+        return smirnoff_data, indexed_mapped_attr_lengths
```

Comment:
WOW. This diff was quite a read. You refactored a lot of complexity. Hopefully the jump to pydantic will crunch this down substantially.

Do you have a materials budget? Is excedrin covered? I'll mail you some if not.

Reply:
Haha, comes with the territory. I keep ibuprofen on hand at all times. :grin: